discovery_engine_core/web-api2/src/models.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{borrow::Cow, collections::HashMap, str::FromStr};
+
+use derive_more::{AsRef, Display};
+use serde::{Deserialize, Serialize};
+
+use xayn_discovery_engine_ai::{Document as AiDocument, Embedding};
+
+use crate::error::common::{InvalidDocumentId, InvalidPropertyId, InvalidUserId};
+
+/// A unique identifier of a document.
+#[derive(
+    AsRef,
+    Clone,
+    Debug,
+    Display,
+    PartialEq,
+    Eq,
+    Hash,
+    Serialize,
+    Deserialize,
+    sqlx::Type,
+    sqlx::FromRow,
+)]
+#[sqlx(transparent)]
+pub struct DocumentId(String);
+
+impl DocumentId {
+    pub fn new(id: impl Into<String>) -> Result<Self, InvalidDocumentId> {
+        let id = id.into();
+
+        if id.is_empty() || id.contains('\u{0000}') {
+            Err(InvalidDocumentId)
+        } else {
+            Ok(Self(id))
+        }
+    }
+
+    pub fn encode(&self) -> Cow<str> {
+        urlencoding::encode(self.as_ref())
+    }
+}
+
+impl FromStr for DocumentId {
+    type Err = InvalidDocumentId;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::new(s)
+    }
+}
+
+impl From<DocumentId> for String {
+    fn from(item: DocumentId) -> Self {
+        item.0
+    }
+}
+
+#[derive(Clone, Debug, Display, Serialize, Deserialize, PartialEq, Eq, Hash, AsRef)]
+pub struct DocumentPropertyId(String);
+
+impl DocumentPropertyId {
+    #[allow(dead_code)]
+    pub fn new(id: impl Into<String>) -> Result<Self, InvalidPropertyId> {
+        let id = id.into();
+
+        if id.is_empty() || id.contains('\u{0000}') {
+            Err(InvalidPropertyId)
+        } else {
+            Ok(Self(id))
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn encode(&self) -> Cow<str> {
```

Comment:
This implies that there is a canonical encoding for this model, while this is only useful when it is used in paths when querying elasticsearch, the model should not be aware of this and in other contexts the required encoding can be different. 
Encoding the id with urlendoding should be done in the elasticsearch code only.

Reply:
I didn't introduce this method here, I moved that code over. 

urlencoding is somewhat of an conancial encoding as it's the only encoding we ever do use and it's not just specific to elastic search but also used when the parameters appear in the path, i.e. it's the encoding _we_ use all over the place, too

I.e. `DocumentId` should implement `FromStr`  (to be usable in `Path` ) later down the line in a way which url decodeds the string. But then we probably also should urlencode it on `Display` as this should match `FromStr`. But then `as_ref()->&str` and `to_string()` would no longer match so we probably should replace `as_ref()` with `as_unencoded_str()` :cry: which doesn't seem nice. Or in other words this is a temporary solution until we decide how to handle the fact that we have two semi-conancial encodings for ids, one is url encoded in path (in our api and elastic search) and the other is not encoded at all in json (in our api and in elastic search).

Butting that aside I really think we should switch to the official elastic search library even if it's in a very early stage, it's still more stable, reliable and well written then any of the ad hoc code we have written for elastic search and using it would safe a bunch of time and we wouldn't have to worry about any encoding of elastic search (we still would have FromStr with url decoding thou, that's in our api).